Add leader election to controller manager

JoelSpeed opened the pull request:
Fixes #93 

This is a first pass at implementing leader election code into the manager.

I've created a `leaderelection` package which contains constructors for a [resourcelock.Interface](https://github.com/kubernetes/client-go/blob/79cb21f5b3b1dd8f8b23bd3f79925b4fda4e2562/tools/leaderelection/resourcelock/interface.go#L57-L76) as well as a fake for use in testing.

I've added three new options to the manager options for leader election, a `bool` to enable the feature (disabled by default to preserve behaviour) and a name and namespace for the configmap that it will use for the lock.

I was considering factoring out the `Start` test suite such that you can run it with different `Options`, since that's all I do in the tests for the leader election and theres a lot of duplication of the test suite. Does that seem like a reasonable thing to do?

DirectXMan12 commented:
> I was considering factoring out ... Does that seem like a reasonable thing to do?

Yeah, it's probably a decent idea.

JoelSpeed commented:
I have refactored the test suite to reduce the duplication.

I asked in Slack in #SIG-api-machinery and got confirmation that configmaps are preferrred over endpoints, however in 1.12 there will be a dedicated object coming for this purpose, so perhaps make a TODO for upgrading it once the controller-runtime gets upgraded to 1.12.

Where would you default the ID and the namespace? I imagined these would be defaulted by adding some flags to wherever the manager is initialised (one of the templates in controller-tools), mine looks like this at the moment:

```
	mgr, err := manager.New(cfg, manager.Options{
		LeaderElection:          *leaderElection,
		LeaderElectionID:        *leaderElectionID,
		LeaderElectionNamespace: *leaederElectionNamespace,
	})
```
Where those variables come from flags.

DirectXMan12 commented:
> TODO for upgrading it once the controller-runtime gets upgraded to 1.12.

Yes, please add the TODO (`// TODO(JoelSpeed): switch to leaderelection object in 1.12`) and/or file an issue

> Where would you default the ID and the namespace

Namespace should default to the current namespace, when we can figure out it (e.g. via some default downward API environment variable).  ID should have some sane default assuming this is the only controller in the namespace, and then we can make it overridable if need be.  We should endeavor to make it so that the defaults work when using the setups that we describe in our docs and in kubebuilder.

JoelSpeed commented:
> Yes, please add the TODO (// TODO(JoelSpeed): switch to leaderelection object in 1.12) and/or file an issue

Added.

> Namespace should default to the current namespace

I've added a `getInClusterNamespace` that attempts to read the namespace from the service account credentials added to a Pod. This should work in most cases I believe. If people aren't mounting service account credentials for their controllers then this won't work however, I can't think when that would be the case though. It returns a "nice" error suggesting you should set the namespace when it isn't running in cluster.

> ID should have some sane default assuming this is the only controller in the namespace

I've defaulted it to `controller-runtime` for now, is about the most sensible thing I could come up with.

DirectXMan12 commented:
I like the defaulting logic, I think this looks decent.  Does the leader election ConfigMap get created with the ID as the name, verbatim?  If so, perhaps something like "controller-leader-election-helper" for the ID.

 :bike: :house:

JoelSpeed commented:
>  Does the leader election ConfigMap get created with the ID as the name, verbatim?

Yes, precisely. If the ID is `controller-runtime` then the ConfigMap would be called `controller-runtime`, I'm happy to change it as per your suggestion, I'll make the change

mhrivnak commented on pkg/leaderelection/leader_election.go:
```diff
@@ -0,0 +1,109 @@
+		return nil, err
+	}
+
+	// TODO(JoelSpeed): switch to leaderelection object in 1.12
```
I'm curious to know more about this but am having a hard time finding other references to it. Do you have any links to info about this object?

JoelSpeed replied:
I asked about it in #sig-api-machinery, and got the info from there: https://kubernetes.slack.com/archives/C0EG7JC6T/p1535318981000100

I know nothing more about it, sorry

mhrivnak replied:
I got this reference from that slack thread: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/coordination/v1beta1/types.go#L27

JoelSpeed commented:
There were merge conflicts so I've just rebased on top of master

JoelSpeed commented:
I fixed the tests that were failing, I think this might be ready to go?

PTAL @DirectXMan12 @pwittrock

DirectXMan12 commented:
Yeah, sorry, trying to do one last scan-through before lgtm.

DirectXMan12 commented:
To be clear, when we lose the leader election lock, we exit out, and presumably the pod gets restarted?  Can you add a note in the `cm.start` or `Start` logic about that?  Otherwise, it's a bit confusing to reason through.  Afterwards, good to go.

JoelSpeed commented:
@DirectXMan12 I've added a note about the behaviour and what to expect when a leader loses leadership. Does that seem sufficient?

DirectXMan12 commented:
Perfect.  Can you squash down into "logical" commits?  We can merge after that.

JoelSpeed commented:
@DirectXMan12 Squashed as per your request 😄

DirectXMan12 commented:
/lgtm
/approve
